<a href="https://colab.research.google.com/github/Aingon14/BSC_DPDM23/blob/main/Frequent_Patterns_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# กลุ่ม ช่วยด้วย
### สมาชิก
        - นางสาวทิพย์วัลย์ สุโพธิ์                 รหัส 643020062-6
        - นางสาวศดานันท์ ทรัพย์มีมหาศาล    รหัส 643021272-0
        - นางสาวอิงอร พลพาล                    รหัส 643021279-6

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/bsc_dpdm23_data'

Mounted at /content/drive


In [3]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [4]:
!pip install apyori  #โหลดเพื่อทำ Association Rule

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=9898169eb913b5577f783534cd9baa9682758a406e86d5b5dcac3f13cec5383b
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [5]:
from apyori import apriori

# HW1 หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลาตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน (groupby ประเทศกับเดือน)

### จัดการวันที่

In [6]:
import datetime

In [7]:
data['InvoiceDate']

0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [8]:
# แปลงคอลัมน์ให้เป็น datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])


In [9]:
#เปลี่ยนรูปแบบให้เป็น เดือน / วัน / ปี
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%y')

In [10]:
# สร้างคอลัมน์ 'Date' โดยดึงข้อมูลเฉพาะวันที่
data['Date'] = data['InvoiceDate'].dt.date
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
...                    ...        ...         ...             ...         ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09  

[541909 rows x 9 columns]

In [11]:
data['month'] = data['InvoiceDate'].dt.month

In [12]:
data['day'] = data['InvoiceDate'].dt.day

# ข้อ 1 หาข้อมูลของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)

In [13]:
# ดูช่วงเวลาคริสต์มาสกับปีใหม่
Christm_Newy = data[data['month']==12]
Christm_Newy

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09   

        month  day  
0          12    1  
1          12    1  
2          12    1  
3          12    1  
4          12    1  
...       ...  ...  
541904     12    9  
541905     12    9  
541906     12    9  
541907     12    9  
541908     12    9  

[68006 rows x 11 columns]

In [14]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data

trans_Christm_Newy = Christm_Newy.groupby('InvoiceNo')['StockCode'].apply(list)

In [15]:
# หาความถี่ของสินค้าที่ถูกซื้อซ้ำๆกันในช่วงเวลาคริสต์มาสกับปีใหม่ โดยมีค่า min support คือ 8%
frequentItemsets= list(apriori(trans_Christm_Newy,min_support = 0.08))
frequentItemsets

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.09769736842105263, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.09769736842105263, lift=1.0)])]

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),

 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),

 RelationRecord(items=frozenset({'85123A'}), support=0.09769736842105263, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.09769736842105263, lift=1.0)])]





In [16]:
# จำนวนสินค้า
len(frequentItemsets)

3

In [17]:
# หาสินค้าที่ถูกซื้อซ้ำจาก trans_Christm_Newy โดยกำหนดค่า min_sup = 0.045  และเช็คว่ามีสินค้ามากกว่า 1 ไอเท็มใน set นั้นหรือไม่

min_sup = 0.045

frequentItem_CN = list(apriori(trans_Christm_Newy,min_support = min_sup))
for i in frequentItem_CN:
    if i[0].__len__()>1:
        print('yes')
        print(i)


yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])


yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0),

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331),

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])

In [18]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22910') | (data['StockCode']=='22086')]['Description'])

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

### สรุปข้อ 1

# ข้อ 2 หาข้อมูลของประเทศในทวีป Europe

In [19]:
#แยกประเทศที่อยู่ในทวีปยุโรป

Europe = [
    'Austria', 'Belgium', 'Channel Islands','Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
    'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy',
    'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden',
    'Switzerland', 'United Kingdom'
]

In [20]:
EU_data = data[data['Country']==""]
for i in range(len(Europe)):
  x = data[data['Country']==Europe[i]]
  EU_data=pd.concat([x,EU_data])
EU_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
536964 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536965 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536966 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536967 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536968 2011-12-08 10:26:00      40.00     12358.0         Austria  2011-12-08   

        month  day  
0          12    1  
1          12    1  
2          12    1  
3          12    1  
4          12    1  
...       ...  ...  
536964     12    8  
536965     12    8  
536966     12    8  
536967     12    8  
536968     12    8  

[538358 rows x 11 columns]

In [21]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data

transactionsEU = EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [22]:
# หาความถี่ของสินค้าที่ถูกซื้อซ้ำๆกันในทวีปยุโรป โดยมีค่า min support คือ 5%

frequentItemsetsEU = list(apriori(transactionsEU,min_support = 0.05))
frequentItemsetsEU

[RelationRecord(items=frozenset({'20725'}), support=0.06187089502934204, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06187089502934204, lift=1.0)]),
 RelationRecord(items=frozenset({'20727'}), support=0.05009521588745094, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20727'}), confidence=0.05009521588745094, lift=1.0)]),
 RelationRecord(items=frozenset({'21212'}), support=0.05149430647856671, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21212'}), confidence=0.05149430647856671, lift=1.0)]),
 RelationRecord(items=frozenset({'22197'}), support=0.05584703276203801, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22197'}), confidence=0.05584703276203801, lift=1.0)]),
 RelationRecord(items=frozenset({'22383'}), support=0.050600443045353855, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({

In [23]:
# จำนวนสินค้า
len(frequentItemsetsEU)

11

In [24]:
# หาสินค้าที่ถูกซื้อซ้ำบ่อยในทวีปยุโรป โดยกำหนดค่า min_sup = 0.07  และเช็คว่ามีสินค้ามากกว่า 1 ไอเท็มใน set นั้นหรือไม่

min_sup = 0.03

frequentItem_EU = list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItem_EU:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItem_EU

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03233453810578679, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03233453810578679, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03011931133652015, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03011931133652015, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7402101241642789, lift=17.20537191045263), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7000903342366757, lift=17

[RelationRecord(items=frozenset({'20712'}), support=0.03365590144184058, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.03365590144184058, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.032567719870972754, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.032567719870972754, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04061249076988846, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04061249076988846, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.06187089502934204, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06187089502934204, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.03999067272939256, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset(

output
[RelationRecord(items=frozenset({'22423'}), support=0.08355679919163654,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08355679919163654, lift=1.0)]),

RelationRecord(items=frozenset({'85099B'}), support=0.0825852085033617, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.0825852085033617, lift=1.0)]),

RelationRecord(items=frozenset({'85123A'}), support=0.08713225292448797, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08713225292448797, lift=1.0)])]

In [25]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22423') | (data['StockCode']=='85099B')| (data['StockCode']=='85123A')]['Description'])

{'?',
 'CREAM HANGING HEART T-LIGHT HOLDER',
 'JUMBO BAG RED RETROSPOT',
 'REGENCY CAKESTAND 3 TIER',
 'WHITE HANGING HEART T-LIGHT HOLDER',
 'damages',
 'faulty',
 'wrongly marked carton 22804'}

## สรุปข้อ 2

#ข้อ3 หาข้อมูลของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)

### จัดการข้อมูลในช่วงตรุษจีนของทวีปเอเชีย

In [27]:
Asia = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [28]:
ASIA_data = data[data['Country']==""]
for i in range(len(Asia)):
  x = data[data['Country']==Asia[i]]
  ASIA_data=pd.concat([x,EU_data])
ASIA_data

InvoiceNo StockCode                         Description  Quantity  \
89570     543911     21485    RETROSPOT HEART HOT WATER BOTTLE         6   
89571     543911     48138                  DOORMAT UNION FLAG         2   
89572     543911     85150       LADIES & GENTLEMEN METAL SIGN        24   
89573     543911     82552                 WASHROOM METAL SIGN        24   
89574     543911     82551              LAUNDRY 15C METAL SIGN        12   
...          ...       ...                                 ...       ...   
536964    581232   15056BL             EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P              EDWARDIAN PARASOL PINK        12   
536966    581232     22059       CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646  CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                             POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID               Country  \
89570  2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571  2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572  2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
...                    ...        ...         ...                   ...   
536964 2011-12-08 10:26:00       5.95     12358.0               Austria   
536965 2011-12-08 10:26:00       5.95     12358.0               Austria   
536966 2011-12-08 10:26:00       0.39     12358.0               Austria   
536967 2011-12-08 10:26:00       0.39     12358.0               Austria   
536968 2011-12-08 10:26:00      40.00     12358.0               Austria   

              Date  month  day  
89570   2011-02-14      2   14  
89571   2011-02-14      2   14  
89572   2011-02-14      2   14  
89573   2011-02-14      2   14  
89574   2011-02-14      2   14  
...            ...    ...  ...  
536964  2011-12-08     12    8  
536965  2011-12-08     12    8  
536966  2011-12-08     12    8  
536967  2011-12-08     12    8  
536968  2011-12-08     12    8  

[538426 rows x 11 columns]

In [32]:
As_data=ASIA_data[(ASIA_data['month']==1)]
As_data.head()

InvoiceNo StockCode                Description  Quantity  \
42481    539993     22386    JUMBO BAG PINK POLKADOT        10   
42482    539993     21499         BLUE POLKADOT WRAP        25   
42483    539993     21498        RED RETROSPOT WRAP         25   
42484    539993     22379   RECYCLING BAG RETROSPOT          5   
42485    539993     20718  RED RETROSPOT SHOPPER BAG        10   

              InvoiceDate  UnitPrice  CustomerID         Country        Date  \
42481 2011-01-04 10:00:00       1.95     13313.0  United Kingdom  2011-01-04   
42482 2011-01-04 10:00:00       0.42     13313.0  United Kingdom  2011-01-04   
42483 2011-01-04 10:00:00       0.42     13313.0  United Kingdom  2011-01-04   
42484 2011-01-04 10:00:00       2.10     13313.0  United Kingdom  2011-01-04   
42485 2011-01-04 10:00:00       1.25     13313.0  United Kingdom  2011-01-04   

       month  day  
42481      1    4  
42482      1    4  
42483      1    4  
42484      1    4  
42485      1    4

In [35]:
As_data_d=ASIA_data[(ASIA_data['day']>=15)]
As_data_d.head()

InvoiceNo StockCode                        Description  Quantity  \
353797    567824      POST                            POSTAGE         1   
31102     538895    90210D      PURPLE ACRYLIC FACETED BANGLE        12   
31103     538895    85180A            RED HEARTS LIGHT CHAIN         12   
31104     538895    85184C  S/4 VALENTINE DECOUPAGE HEART BOX        24   
31105     538895     84826     ASSTD DESIGN 3D PAPER STICKERS        60   

               InvoiceDate  UnitPrice  CustomerID               Country  \
353797 2011-09-22 13:00:00      37.50     12739.0  United Arab Emirates   
31102  2010-12-15 09:07:00       1.25     14401.0        United Kingdom   
31103  2010-12-15 09:07:00       1.25     14401.0        United Kingdom   
31104  2010-12-15 09:07:00       1.25     14401.0        United Kingdom   
31105  2010-12-15 09:07:00       0.21     14401.0        United Kingdom   

              Date  month  day  
353797  2011-09-22      9   22  
31102   2010-12-15     12   15  
31103   2010-12-15     12   15  
31104   2010-12-15     12   15  
31105   2010-12-15     12   15

In [40]:
As_data2=ASIA_data[(ASIA_data['month']==2)]
As_data2.head()

InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571 2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572 2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   

             Date  month  day  
89570  2011-02-14      2   14  
89571  2011-02-14      2   14  
89572  2011-02-14      2   14  
89573  2011-02-14      2   14  
89574  2011-02-14      2   14

In [38]:
As_data_d2=ASIA_data[(ASIA_data['day']<=15)]
As_data_d.head()

InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571 2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572 2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   

             Date  month  day  
89570  2011-02-14      2   14  
89571  2011-02-14      2   14  
89572  2011-02-14      2   14  
89573  2011-02-14      2   14  
89574  2011-02-14      2   14

In [42]:
# รวมทั้ง 2 เดือนเข้าด้วยกัน ในช่วงตรุษจีน
Asia_data_chi=pd.concat([As_data_d,As_data_d2])
Asia_data_chi

InvoiceNo StockCode                         Description  Quantity  \
89570     543911     21485    RETROSPOT HEART HOT WATER BOTTLE         6   
89571     543911     48138                  DOORMAT UNION FLAG         2   
89572     543911     85150       LADIES & GENTLEMEN METAL SIGN        24   
89573     543911     82552                 WASHROOM METAL SIGN        24   
89574     543911     82551              LAUNDRY 15C METAL SIGN        12   
...          ...       ...                                 ...       ...   
536964    581232   15056BL             EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P              EDWARDIAN PARASOL PINK        12   
536966    581232     22059       CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646  CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                             POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID               Country  \
89570  2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571  2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572  2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
...                    ...        ...         ...                   ...   
536964 2011-12-08 10:26:00       5.95     12358.0               Austria   
536965 2011-12-08 10:26:00       5.95     12358.0               Austria   
536966 2011-12-08 10:26:00       0.39     12358.0               Austria   
536967 2011-12-08 10:26:00       0.39     12358.0               Austria   
536968 2011-12-08 10:26:00      40.00     12358.0               Austria   

              Date  month  day  
89570   2011-02-14      2   14  
89571   2011-02-14      2   14  
89572   2011-02-14      2   14  
89573   2011-02-14      2   14  
89574   2011-02-14      2   14  
...            ...    ...  ...  
536964  2011-12-08     12    8  
536965  2011-12-08     12    8  
536966  2011-12-08     12    8  
536967  2011-12-08     12    8  
536968  2011-12-08     12    8  

[566180 rows x 11 columns]

### หาข้อมูล

In [ ]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data
transactionsAsiaChi = Asia_data_chi.groupby('InvoiceNo')['StockCode'].apply(list)

In [52]:
min_sup = 0.08
frequentItemsets_Asia= list(apriori(transactionsAsiaChi,min_support = min_sup))
for i in frequentItemsets_Asia:
    if i[0].__len__()>3:
        print('yes')
        print(i)
frequentItemsets_Asia

[RelationRecord(items=frozenset({'22423'}), support=0.08432400502104408, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08432400502104408, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.0823303551650299, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.0823303551650299, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08993576017130621, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08993576017130621, lift=1.0)])]

[RelationRecord(items=frozenset({'22423'}), support=0.08432400502104408, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08432400502104408, lift=1.0)]),

 RelationRecord(items=frozenset({'85099B'}), support=0.0823303551650299, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.0823303551650299, lift=1.0)]),


 RelationRecord(items=frozenset({'85123A'}), support=0.08993576017130621, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08993576017130621, lift=1.0)])]

In [53]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22423') | (data['StockCode']=='85099B')| (data['StockCode']=='85123A')]['Description'])

{'?',
 'CREAM HANGING HEART T-LIGHT HOLDER',
 'JUMBO BAG RED RETROSPOT',
 'REGENCY CAKESTAND 3 TIER',
 'WHITE HANGING HEART T-LIGHT HOLDER',
 'damages',
 'faulty',
 'wrongly marked carton 22804'}

## สรุปข้อ 3